src/test/scala/chiselTests/LoadMemoryFromFileSpec.scala

```diff
@@ -0,0 +1,107 @@
+    val value2  = Output(memoryType)
+  })
+
+  val memory = Mem(memoryDepth, memoryType)
+
+  chisel3.experimental.annotate(ChiselLoadMemoryAnnotation(memory, "./mem1"))
+  io.value := memory(io.address)
+
+  val low = Module(new UsesMemLow(memoryDepth, memoryType))
+
+  low.io.address := io.address
+  io.value2 := low.io.value
+}
+
+class UsesMemLow(memoryDepth: Int, memoryType: Data) extends Module {
+  val io = IO(new Bundle {
+    val address = Input(UInt(memoryType.getWidth.W))
+    val value   = Output(memoryType)
+  })
+
+  val memory = Mem(memoryDepth, memoryType)
+
+  chisel3.experimental.annotate(ChiselLoadMemoryAnnotation(memory, "./mem2"))
+  io.value := memory(io.address)
+}
+
+//noinspection TypeAnnotation
+class FileHasSuffix(memoryDepth: Int, memoryType: Data) extends Module {
+  val io = IO(new Bundle {
+    val address = Input(UInt(memoryType.getWidth.W))
+    val value   = Output(memoryType)
+    val value2  = Output(memoryType)
+  })
+
+  val memory = Mem(memoryDepth, memoryType)
+
+  chisel3.experimental.annotate(ChiselLoadMemoryAnnotation(memory, "./mem1.txt"))
+  io.value := memory(io.address)
+
+  val low = Module(new UsesMemLow(memoryDepth, memoryType))
+
+  low.io.address := io.address
+  io.value2 := low.io.value
+}
+
+class LoadMemoryFromFileSpec extends FreeSpec with Matchers {
+  val testDirName = "test_run_dir/load_memory_spec"
+  "Users can specify a source file to load memory from" in {
+    val result = Driver.execute(
+      args = Array("-X", "verilog", "--target-dir", testDirName),
+      dut = () => new UsesMem(memoryDepth = 8, memoryType = UInt(16.W))
+    )
+
+    result match {
+      case ChiselExecutionSuccess(_, _, Some(FirrtlExecutionSuccess(_, _))) =>
+        val dir = new File(testDirName)
+        new File(dir, "UsesMem.UsesMem.memory.v").exists() should be (true)
+        new File(dir, "UsesMem.UsesMemLow.memory.v").exists() should be (true)
+        new File(dir, "firrtl_black_box_resource_files.f").exists() should be (true)
+      case _=>
+        throw new Exception("Failed compile")
+    }
+  }
+
+  "Load memory annotations should not have suffixes" in {
+    val testDirName = "test_run_dir/load_memory_suffix_spec"
+
+    val outputBuffer = new ByteArrayOutputStream()
+    Console.withOut(new PrintStream(outputBuffer)) {
+      val result = Driver.execute(
+        args = Array("-X", "verilog", "--target-dir", testDirName),
+        dut = () => new FileHasSuffix(memoryDepth = 8, memoryType = UInt(16.W))
+      )
+
+      result match {
+        case ChiselExecutionSuccess(_, _, Some(FirrtlExecutionSuccess(_, _))) =>
+          val dir = new File(testDirName)
+
+        case _ =>
+          throw new Exception("Failed compile")
+      }
+    }
+    val output = outputBuffer.toString
+
+    output should include (
+      s"""LoadMemoryAnnotation fileName "./mem1.txt" has extension ".txt" will still be appended"""
+    )
+    output should include ("""1 warning(s)""")
+  }
+
+}
```

Comment:
I'd like to see this PR add some end-to-end testing of memories. I.e., add some resources of readmemh/readmemb format and then make sure that the data matches up.

Reply:
I have end to end testing in an as-yet-unpushed PR on chisel-testers because of all the heavy lifting that is done there to invoke verilator with all the right arguments and file names. Because of the testers2 project Richard is working on I'm not really sure where that is all going to live come 3.2, but I think it is as yet un-PR'd chisel3

Reply:
Sounds great. I should have expected that you had something lined up...